Add Components page to developer guide (#1865)
MarkBind/markbind: docs/devGuide/writingComponents.md

{% set title = "Writing Components" %}
<span id="title" class="d-none">{{ title }}</span>

<frontmatter>
  title: "{{ title }}"
  layout: devGuide.md
  pageNav: default
</frontmatter>

# {{ title }}

<div class="lead">

This page explains various concerns related to MarkBind components, focused on implementation and testing.
</div>

> 

Reviewer: I think we can include a quick introduction here of 
1. what 'components' refer to & what a 'node' is.
1. since you use the words slot and attribute later on, maybe we can put in a quick description of the meaning now? In case people read it without reading through the whole component guide or need a refresher
1. The node processing workflow (explained in paragraphs 2 and 3  under transforming the node directly), since it seems applicable to all components and not just that part?

Author: Thanks for the suggestion! I think explaining node processing is a good idea since it's pretty important, will add it to the guide :o

Reviewer: Yay, really helpful introduction!

## Implementing Components

There are multiple ways to implement MarkBind components.

> <box type='warning' light>

Reviewer: I just noticed that most of the devGuide uses the seamless style for boxes! I think it might be good to use them on this page too, just for visual consistency

If an author has a conflicting slot and attribute, MarkBind should **log a warning** to let them know!

> </box>

Author: I agree, thanks for the suggestion! Added a new slots/attributes section at the end to address this (and related concerns)

These transformations may take place at various stages of node processing: before (`preProcessNode`), during (`processNode`), or after (`postProcessNode`).

{{ icon_examples }} 
* Adding a class to a node (setting line numbers for code blocks)
* Modifying attributes and adding a directive to a node ([former implementation](https://github.com/MarkBind/markbind/blob/502df135e07baebd9d4eea8ccc0654c990047792/packages/core/src/html/bootstrapVueProcessor.js#L73) of popovers)

</box>

### Vue Components

Many MarkBind components are implemented as Vue components, either by creating a component in the `vue-components` package, or by importing a component from an external library.
This can be useful when a more complicated set of features is needed, where a Vue component can provide an interface for us to manage these functionalities.

Vue components are registered in `vue-components/src/index.js`, which allows them to be used in the template section of any Vue instance without needing to be imported first.

<panel header="How do MarkBind attributes/slots get passed to the Vue component?">

> ##### Attributes

Reviewer: I think we can include the same tooltip information here as you did below; the reader won't see the bottom definition until after they read this if they read chronologically

MarkBind attributes are passed to the Vue component as **props**. The type of the prop will be a `String`.

##### Slots

MarkBind slots are passed as **named slots** to the Vue component. The name of the MarkBind slot will be the same as the name of the Vue slot.
Hence, MarkBind slots can be accessed in a Vue component either through the [named slots](https://v2.vuejs.org/v2/guide/components-slots.html#Named-Slots) or through the [`$slots` API](https://v2.vuejs.org/v2/api/#vm-slots).
</panel>

<br>

{{ icon_examples }} 
* As a wrapper for an external library (Modal component)
* To implement a set of customised behaviours (Quiz component)

### As a Plugin

MarkBind components can be implemented as a plugin as well. 

> This is suitable for more lightweight components where the implementation is largely in processing the node, where it is fitting to use MarkBind Plugins' `processNode` or `postRender` interfaces.

Reviewer: Hmm, I think the difference between a plugin and transforming the node directly is not clear from the documentation

{{ icon_examples }} 
* The `tree` component is implemented as a default plugin

## Testing Components

Automated tests that are relevant to the components include:

* [Functional tests]({{baseUrl}}/devGuide/workflow.html#adding-test-site-content)
* [Snapshot tests]({{baseUrl}}/devGuide/workflow.html#adding-snapshot-tests-for-components)

The API for Snapshot tests can be found at [Vue Test Utils](https://v1.test-utils.vuejs.org/).

Additionally, it's a good idea to check the deployed PR preview in addition to serving the app locally, to ensure that there are no differences.

## Additional Considerations

Some things you may need to consider when implementing a MarkBind component:

#### Reactivity

_Reactivity_ refers to the ability of a web framework to update your view whenever the application state has changed. 
It is important to consider reactivity when implementing a component that may have dynamic contents that readers can interact with (e.g. opening a panel, triggering a tooltip to show).

> #### SSR

Reviewer: Can we include what SSR stands for here (Server-side Rendering)?

Components should be compatible with SSR. 
Minimally, there should be no SSR issues (viewable from the browser console), though a lack of warnings does **not** mean that there are no SSR problems. 
A guide on SSR for MarkBind can be found [here]({{baseUrl}}/devGuide/serverSideRendering.html). 

Vue-component-specific tips for resolving SSR issues:
* The `mount` and `beforeMount` lifecycle hooks will only be executed on the client, not the server
* When using `v-if`, ensure that it will evaluate to the same value on both the client and server 
* Take note of how the Vue component will be compiled, ensure that the HTML is correct and aligns on both client- and server- side
* Conditionally render data when it has been fully loaded

#### Bundle size

When creating a new component, you may need to import a package or library to support some functionality. 
Ideally, this should not increase MarkBind's bundle size too much.

#### Dependencies

> When choosing to use an external library, it should ideally be well-maintained and not have too many dependencies, especially high-level dependencies. High-level dependencies may lag behind the most recent releases of other libraries, which may become a blocker for MarkBind to migrate to these recent releases as well.

Reviewer 2: Perhaps can add a note to encourage the developer to propose/raise this during the initial discussion phase before proceeding to importing/adding a dependency

{% set title = "Writing Components" %}
<span id="title" class="d-none">{{ title }}</span>

<frontmatter>
  title: "{{ title }}"
  layout: devGuide.md
  pageNav: default
</frontmatter>

# {{ title }}

<div class="lead">

This page explains how MarkBind components work, focused on implementation and testing.
</div>

> MarkBind provides a number of components (e.g. expandable panels, tooltips) dynamically express content. 

Reviewer 2: ```suggestion
MarkBind provides a number of components (e.g. expandable panels, tooltips) that dynamically express content. 
```

In order to serve content on the browser, MarkBind syntax is converted to valid HTML.

<panel header="How are components in MarkBind syntax parsed and converted to HTML?">

> The main logic of the node processing flow can be found in `NodeProcessor`.

Reviewer 3: Maybe can include a link to the `NodeProcessor` file? (Or alternatively can state the directory path of the file so that it is easier for the readers to find and reference the file.)

Author: Added both since MarkBind link formatting doesn't show up well on `inline code`!

Many MarkBind components are implemented as Vue components, either by creating a component in the `vue-components` package, or by importing a component from an external library.
This can be useful when a more complicated set of features is needed, where a Vue component can provide an interface for us to manage these functionalities.

Vue components are registered in `vue-components/src/index.js`, which allows them to be used in the template section of any Vue instance without needing to be imported first.

<panel header="How do MarkBind attributes/slots get passed to the Vue component?">

##### Attributes

MarkBind attributes are passed to the Vue component as **props**. The type of the prop will be a `String`.

##### Slots

MarkBind slots are passed as **named slots** to the Vue component. The name of the MarkBind slot will be the same as the name of the Vue slot.
Hence, MarkBind slots can be accessed in a Vue component either through the [named slots](https://v2.vuejs.org/v2/guide/components-slots.html#Named-Slots) or through the [`$slots` API](https://v2.vuejs.org/v2/api/#vm-slots).
</panel>

<br>

{{ icon_examples }} 
* As a wrapper for an external library (Modal component)

> * To implement a set of customised behaviours (Quiz component)

Reviewer 3: Similarly, can either link these to the files or state their paths :)

### As a Plugin

MarkBind components can be implemented as a plugin as well. 
This is suitable for more lightweight components where the implementation is largely in processing the node, where it is fitting to use MarkBind Plugins' `processNode` or `postRender` interfaces.

{{ icon_examples }} 

> * The `tree` component is implemented as a default plugin

Reviewer 3: Same here :)

_Reactivity_ refers to the ability of a web framework to update your view whenever the application state has changed. 
It is important to consider reactivity when implementing a component that may have dynamic contents that readers can interact with (e.g. opening a panel, triggering a tooltip to show).

#### SSR

Components should be compatible with SSR. 
Minimally, there should be no SSR issues (viewable from the browser console), though a lack of warnings does **not** mean that there are no SSR problems. 
A guide on SSR for MarkBind can be found [here]({{baseUrl}}/devGuide/design/serverSideRendering.html). 

Vue-component-specific tips for resolving SSR issues:
* The `mount` and `beforeMount` lifecycle hooks will only be executed on the client, not the server
* When using `v-if`, ensure that it will evaluate to the same value on both the client and server 
* Take note of how the Vue component will be compiled, ensure that the HTML is correct and aligns on both client- and server- side
* Conditionally render data when it has been fully loaded

#### Bundle size

When creating a new component, you may need to import a package or library to support some functionality. 

> Ideally, this should not increase MarkBind's bundle size too much.

Reviewer 3: Maybe can provide an example of how the bundle size can be checked? I think [bundlephobia](https://bundlephobia.com/) is quite a good site to check the size of a package/library but feel free to put in other methods if there are better alternatives :)

Author: Thanks for the suggestion! I had no idea that existed :o

#### Dependencies

When choosing to use a third-party library or package, it should ideally be well-maintained and not have too many dependencies, especially high-level dependencies. 

> While dependencies may be inevitable, a package that depends on <tooltip content="e.g. frameworks like Vue, Bootstrap">higher-level libraries</tooltip> may lag behind the most recent releases of these libraries, which may become a blocker for MarkBind to migrate to these recent releases as well.

Reviewer 2: ```suggestion
While dependencies may be inevitable, a package that depends on <tooltip content="e.g. frameworks like Vue, Bootstrap">lower-level libraries</tooltip> may lag behind the most recent releases of these libraries, which may become a blocker for MarkBind to migrate to these recent releases as well.
```

Is this a typo?

Reviewer 2: I was thinking of BootStrap-Vue (higher level) vs Vue(lower level) when reading this. I think an example of what higher/lower means would be nice, or just using an example would be fine as well.

Author: Ah ok, in that case I have removed mentions of high- vs low- level since it may be quite relative depending on the context. Replaced with an example on bootstrap-vue!

{% set title = "Writing Components" %}
<span id="title" class="d-none">{{ title }}</span>

<frontmatter>
  title: "{{ title }}"
  layout: devGuide.md
  pageNav: default
</frontmatter>

# {{ title }}

<div class="lead">

This page explains how MarkBind components work, focused on implementation and testing.
</div>

MarkBind provides a number of components (e.g. expandable panels, tooltips) to dynamically express content. 
In order to serve content on the browser, MarkBind syntax is converted to valid HTML.

<panel header="How are components in MarkBind syntax parsed and converted to HTML?">

The main logic of the node processing flow can be found in [`packages/core/src/html/NodeProcessor.js`](https://github.com/MarkBind/markbind/blob/master/packages/core/src/html/NodeProcessor.js).

A MarkBind source file is first parsed into a series of 
<popover header=":bulb: What is a _**node**_?" content="A HTML file can be represented as a tree structure called the [DOM](https://www.w3schools.com/js/js_htmldom.asp), comprising HTML elements (or _nodes_).">nodes</popover>.

> In general, each component will be parsed as a node, which may contain other childnodes (components or otherwise).

Reviewer 2: ```suggestion
In general, each component will be parsed as a node, which may contain other child nodes (components or otherwise).
```
